### ISSUE TITLE
Shouldn't `late` fields be exempted from `overridden_fields` lint?

### ISSUE THREAD

**Issue by AlexanderFarkas** [dart-lang/language#3421]
[Lint rule](https://dart.dev/tools/linter-rules/overridden_fields)

I believe, the only concern is having both `super.field` and `this.field` initialized without `super.field` being used or garbage collected. But if they are defined as `late`, I don't see a reason to lint it.

The following:
```dart
class A {
  final number = () {
    print("A init");
    return 1;
  }();
  
  
  void printNumber() {
    print(number);
  }
}

class B extends A {
  final number = () {
    print("B init");
    return 2;
  }();
}


void main() {
  final b = B();
  b.printNumber();
}
```

Prints:
```shell
B init
A init
2
```

But, if I change `number` in both classes to `late`, like that:
```dart
class A {
  late final number = () {
    print("A init");
    return 1;
  }();
  
  
  void printNumber() {
    print(number);
  }
}

class B extends A {
  late final number = () {
    print("B init");
    return 2;
  }();
}
```

It prints only:
```
B init
2
```

**Comment by AlexanderFarkas** [comment]
To clarify, the issue is not about lint itself.
It's about the reasoning behind it. 

So, I believe the issue should be moved to SDK repository only if `late final` fields might be safely overridden.

**Comment by lrhn** [comment]
There is no inherent difference between late and non-late fields that matters here.

Both represent a memory slot in the object, which may be wasted if it's never accessed. That is, if the program contains no reachable read of the field's getter.

A clever compiler may be able to recognize that, and optimize the slot away (and if overridden with another field, maybe reuse the slot).
And if that falls, the slot is still there and wastes memory.

If anything, a late field may use a bit more memory, to remember whether it's initialized.

**Comment by AlexanderFarkas** [comment]
@lrhn Thank you for the response! 

If I understand correctly, overridden late fields only allocate slot and bool (for initialization status) - they don't allocate memory for its value.

So, overriding late fields results in loss of several bytes - with no side effects 

Am I correct or not?

**Comment by lrhn** [comment]
Correct, but the same applies to a non-late nullable field containing `null`, or integer field containing a small integer, those also don't use memory on the value. The lint still applies to those, because it is really about the unnecessary memory slot.

**Comment by eernstg** [comment]
> If I understand correctly, overridden late fields only allocate slot and bool (for initialization status) - they don't allocate memory for its value.

Hi @AlexanderFarkas, I'll add another comment to make sure we don't leave a possible misunderstanding hanging in the air.

A late non-local variable (any of them: instance variable, static variable, top-level variable) is supported by storage that holds the value plus possibly storage that holds the initialization status. (With a non-nullable declared type we can use the value storage holding null to indicate that the variable is not yet initialized, because that could never be the value of the variable.) As @lrhn mentioned, there might be an opportunity for a smart compiler to omit one or two of these slots for a late instance variable in an override relation (it could be the overridden one or the overriding one), but there could be many reasons why this optimization can't be applied.

So we're much better off if we just never override a variable by another variable (`late` or not, that doesn't matter). To achieve that, we could change the overridden declaration (say, `int i;` becomes `abstract int i;` or `int get i;`) or the superinterface relation (say, `extends C` becomes `implements C` so that we don't inherit that variable from `C`), or we could use any of several other techniques.

In any case, I can't a situation where it is true that 'overridden late fields ... don't allocate memory for its value'&mdash;they certainly do!
